persistence-test/src/main/java/io/stargate/it/PersistenceTest.java

```diff
@@ -1729,6 +1733,86 @@ public void testPagingStateForNextPartition() throws ExecutionException, Interru
     assertThat(rs.rows()).hasSize(2);
   }
 
+  @Test
+  public void testPagingStateForNextRow() throws ExecutionException, InterruptedException {
+    setupCustomPagingData();
+
+    // Obtain partition keys in "ring" order
+    AbstractBound<?> selectAll =
+        dataStore
+            .queryBuilder()
+            .select()
+            .column("pk")
+            .column("val")
+            .from(keyspace, table)
+            .build()
+            .bind();
+
+    java.util.List<Row> allRows = dataStore.execute(selectAll).get().rows();
+    List<Integer> keys = allRows.stream().map(r -> r.getInt(0)).collect(Collectors.toList());
+    List<Integer> values = allRows.stream().map(r -> r.getInt(1)).collect(Collectors.toList());
+    int key2 = keys.get(1);
+    int val2 = values.get(1);
+    int key3 = keys.get(2);
+    int val3 = values.get(2);
+    int key6 = keys.get(5);
+    int val6 = values.get(5);
+    int key8 = keys.get(8);
+    int val8 = values.get(8);
+
+    AbstractBound<?> select =
+        dataStore
+            .queryBuilder()
+            .select()
+            .column("pk")
+            .column("cc")
+            .column("val")
+            .from(keyspace, table)
+            .perPartitionLimit(3)
+            .limit(10)
+            .build()
+            .bind();
+
+    ResultSet rs = dataStore.execute(select, p -> p.toBuilder().pageSize(5).build()).get();
+    java.util.List<Row> rows = rs.currentPageRows();
+    assertThat(rows).hasSize(5); // full page
+    Row row2 = rows.get(1);
+    assertThat(row2.getInt(0)).isEqualTo(key2);
+    assertThat(row2.getInt(1)).isEqualTo(val2);
+    // with 3 row per partition, the last row in page 1 is the second row of the second partition
```

Comment:
I think that this test will be more readable if it will be split into N test cases. (or parameterized test)

Reply:
I'm not sure it can be split. It's really just one case with two reads. The first read gets some data and the second read uses a custom paging position to re-fetch some of the row from the first read, essentially back-tracking the pagination a bit. Information is shared between these operations from the test case perspective.